code/modules/tools/tool_atom.dm

```diff
@@ -0,0 +1,39 @@
+/atom/proc/get_tool_quality(var/archetype)
+	var/datum/extension/tool/tool = get_extension(src, /datum/extension/tool)
+	. = tool?.get_tool_quality(archetype)
+
+/atom/proc/get_tool_speed(var/archetype)
+	var/datum/extension/tool/tool = get_extension(src, /datum/extension/tool)
+	. = tool?.get_tool_speed(archetype)
+
+/atom/proc/do_tool_interaction(var/archetype, var/mob/user, var/atom/target, var/delay = (1 SECOND), var/start_message, var/success_message, var/failure_message, var/fuel_expenditure = 0)
+
+	if(get_tool_quality(archetype) <= 0)
+		return FALSE
+
+	if(start_message)
+		to_chat(user, start_message)
+
+	var/datum/extension/tool/tool = get_extension(src, /datum/extension/tool)
+	. = tool.do_tool_interaction(archetype, user, target, delay, fuel_expenditure)
+
+	if(!QDELETED(user))
+		if(.)
+			if(success_message)
+				to_chat(user, success_message)
+		else
+			if(failure_message)
+				to_chat(user, failure_message)
```

Comment:
This, on the other hand, seems like it will be double-messaging if supplied, as you already give some error messages in the action proc. Maybe either pass the `success_message` and `fail_message` parameters into the `tool.do_tool_interaction` step for more consistent handling, or double-check that what I am saying is false.

Reply:
Fixed.